### PR TITLE
Cycle detection solely using the Dep_stack

### PR DESCRIPTION
**Pull request by rgrinberg**
@diml this PR implements the cycle detection change we discussed last week. I first implemented this in #2930, but then realized it can be extracted into its own PR.

### CODE REVIEW

**Comment by rgrinberg** [comment]
Since the cycle detection is a bit more eager, we currently lose this part of the error message:

```
  -> required by library "c" in _build/default
  -> required by executable cycle in dune:17
```

To completely restore the old errors, `instantiate` needs to return `lib Or_exn.t` rather than `lib` like it did before.

**Comment by aalekseyev** [comment]
I guess this changes behavior in some cases:
(correct me if I'm wrong)
1. before this PR, dependency cycles in optional deps libraries are not an error: the library is just dropped.
2. before this PR, if you don't use a library with a dependency cycle then there is no error.

I think we should add tests for these behaviors and a changelog entry. (or I'm just confused)

**Comment by rgrinberg** [comment]
No, you got it right. Are you convinced that 2 is undesirable? It seems like users can keep incorrect library definitions in their workspace by accident. On the other hand, introducing a cycle mid-refactoring does seem like it could be useful.

**Comment by aalekseyev** [comment]
I think (2) is undesirable, but actually I'm not sure that's changing in this PR.

One way to think about the desirability of (2) is to consider compositionality: adding/removing a project to a workspace shouldn't affect the build outcomes of unrelated projects.

Another angle is performance: if dune is asked to build base, it shouldn't spend time exploring the entire world looking for cycles.

(that said, neither of these reasons is particularly strong)

**Comment by rgrinberg** [comment]
> Another angle is performance: if dune is asked to build base, it shouldn't spend time exploring the entire world looking for cycles.

In theory, this shouldn't matter though. Dune shouldn't ask for to resolve a library name that isn't going to be used. In practice, I know that dune is a bit flawed in this department.

In any case, I will try to restore the old behavior. I wasn't trying to change neither (1) nor (2).

**Comment by rgrinberg** [comment]
@aalekseyev I restored the old behavior.

Regarding optional deps, do you mind if I address this in another PR? It's a bit tougher than I expected, and perhaps will need a different approach. The gist of it is that optional deps should be hidden when their dependencies are possibly missing. Currently, our check for unavailable dependencies is far too wide. However, it's not entirely clear to me how to improve in this situation:

```
(library (name y) (libraries foo))
(library
 (name foo)
 (optional)
 (libraries x y))
```

Where `x` is missing. In this case, we know this library is at least `optional` after analyzing `x`. Should we keep analyzing the other deps of `foo` to check for cycles?

**Comment by aalekseyev** [comment]
Could you add that example as a test case? I'm not sure at all what the best behavior is, but in any case it's good to understand what it is and how it changed.

File: src/dune/lib.ml [hunk 1]
```diff
@@ -994,10 +1024,10 @@ end = struct
     Option.iter (Table.find db.table name) ~f:(fun x ->
         already_in_table info name x);
     (* Add [id] to the table, to detect loops *)
```
Comment (aalekseyev) [inline src/dune/lib.ml]: This comment is stale.

**Comment by aalekseyev** [comment]
Sorry, this is a concern I should have raised at the very beginning, but it only crystallized now...

The main change in this feature is removal of blackholes (`St_initializing`). Indeed cycle detection is one purpose serve, but there's another: it's making sure that we only resolve each library once.

As it stands, I can't convince myself that `Table.set db.table name res` on line 1158 won't happen twice for the same library (and if it does, maybe it breaks things e.g. because `unique_id` is different).

**Comment by rgrinberg** [comment]
> The main change in this feature is removal of blackholes (St_initializing). Indeed cycle detection is one purpose serve, but there's another: it's making sure that we only resolve each library once.

Well I have an upcoming PR to change the resolution to be independent of instantiation. The idea would be that:

```

val resolve : db -> Lib_name.t -> Id.t * Lib_info.t
```

would be memoized. I think that should certainly guarantee that resolution will only be done once.

> As it stands, I can't convince myself that Table.set db.table name res on line 1158 won't happen twice for the same library (and if it does, maybe it breaks things e.g. because unique_id is different).

Indeed, it would be safer if it was `add_exn`. I made that change, thanks.

I think the current code is as safe as the old one in the sense:

* The behavior for correct builds is exactly the same
* The behavior for incorrect builds is likely the same, but the error message might *possibly* be different. Like you, i'm not fully convinced that instantiate will not hit one of those code errors on a cycle.

For me, it would be easiest to improve the state of this code by making additional PR's rather than packing everything here. However, if you are uneasy about it, I can try to do it all here too.

**Comment by aalekseyev** [comment]
> For me, it would be easiest to improve the state of this code by making additional PR's rather than packing everything here. However, if you are uneasy about it, I can try to do it all here too.

In general I think leaving master in a suspicious state is not great, but in this case I think `add_exn` solves it completely. Thanks.

**Comment by aalekseyev** [comment]
Oh well, looks like the tests are failing now.

**Pull request closed without merging**